[wip, test CI] Add IterableDataset

ssnl opened the pull request:
1. Add `IterableDataset`.
2. Support non batched loading of traditional map-like dataset. This is useful in doing bulk loading.
3. So we have three data loader mods: `Iterable` (newly added), `Map` (newly added), and `MapWithBatchedRead` (old).

    1. `Iterable` if the `dataset` is an instance of `IterableDataset`
    2. `Map` if `batch_size` is `None`
    3. `MapWithBatchedRead` is chosen otherwise.

3. Refactor `DataLoaderIter` into two classes, `_SingleProcessDataLoaderIter` and `_MultiProcessingDataLoaderIter`. Rename some methods to be more generic, e.g., `get_batch` -> `get_data`.
4. Add `torch.utils.data.get_worker_info` which returns worker information in a worker proc (e.g., worker id, dataset obj copy, etc.) and can be used in `IterableDataset.__iter__` and `worker_init_fn` to do per-worker configuration.
5. Add `ChainDataset`, which is the analog of `ConcatDataset` for `IterableDataset`.
6. Add `convert_fn`, which is meant to convert each loaded data into tensors. For the `MapWithBatchedRead` mode, fetched data is first converted using `convert_fn` and then collated. This shouldn't be much slower (if any) than the old approach of only using `collate_fn`.

ssnl committed:
Add IterableDataset

ssnl committed:
Add torch.utils.data.get_worker_info

ssnl committed:
test_proper_exit prints more informative error message

kittipatv commented on torch/utils/data/dataset.py:
```diff
@@ -23,8 +24,26 @@ def __add__(self, other):
+    r"""An iterable Dataset.
+
+    All other datasets should subclass it. All subclasses should override
+    ``__len__``, that provides the size of the dataset, and ``__getitem__``,
```
> All subclasses should override ``__len__``

Is this a hard requirement? What if we have a stream of data?

SsnL replied:
No it's not. I just copied from Dataset and haven't changed the doc. :) As long as you overwrite `__iter__`, you are good.

apaszke commented:
Can you please elaborate on the final design of iterable datasets? As I've mentioned in some PRs to C++ Dataset API, I think that our current DataLoader story is completely mismatched with the desire to have mutable datasets (including iterable datasets). In particular, you can't simply replicate the dataset, and have all workers produce the same samples, because that would be silly. Similarly, the Sampler is completely useless in this case. I really think we should work on alternative solutions for this API.

SsnL commented on test/test_dataloader.py:
```diff
@@ -231,30 +231,50 @@ def __len__(self):
+    def __init__(self, sizes_for_all_workers):
+        self.sizes_for_all_workers = sizes_for_all_workers
+
+    def __iter__(self):
```
@apaszke Basically you can config each dataset replica differently in two different ways using `torch.utils.data.get_worker_info()` which returns the worker's `id`, `seed` and `dataset` replica:
1. In `__iter__` for an iterable dataset. 
2. In `worker_init_fn`.

This requires user to write their dataset code with multiprocessing data loading in mind. But I think this is a reasonable requirement because (1) there is no general way to split an iterator across multiple processes (2) if people want to do sharding / bulk loading, how to split work among workers should be part of the design.

ssnl committed:
fix default_collate

ssnl committed:
Fix single process loading, Add check in ConcatDataset, Add tests

ssnl committed:
add __len__ to tested iterable dataset to support zip in py2

apaszke reviewed (commented):
Haven't read the new multiprocess iterator, but I had a few questions

SsnL commented on torch/utils/data/_utils/worker.py:
```diff
@@ -55,7 +56,46 @@ def is_alive(self):
 
 
-def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed, init_fn, worker_id):
+WorkerInfo = namedtuple('WorkerInfo', ['id', 'seed', 'dataset'])
```
Yep I agree. Will make the change.

SsnL commented on torch/utils/data/_utils/collate.py:
```diff
@@ -9,60 +9,89 @@
 import re
 from torch._six import container_abcs, string_classes, int_classes
 
+np_str_obj_array_pattern = re.compile(r'[SaUO]')
```
lol before #14668  we were using `re.search`, which is worse.

SsnL commented on torch/utils/data/_utils/collate.py:
```diff
@@ -9,60 +9,89 @@
+                    strides.insert(0, strides[0] * s)
+                out = out.set_(out.storage, out_shape, strides)
+            for src, dst in zip(batch, out):
+                dst.copy_(src)
```
sure

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -70,47 +79,61 @@ class DataLoader(object):
     def __init__(self, dataset, batch_size=1, shuffle=False, sampler=None,
-                 batch_sampler=None, num_workers=0, collate_fn=default_collate,
+                 batch_sampler=None, num_workers=0,
+                 convert_fn=_utils.collate.default_convert,
```
With `collate_fn`, the input is assumed to be a list of **many data samples**, and you want to collate each field in the data structure of each element. E.g., `[([1, 2], [3, 4]), ([5, 6], [7, 8)]` should be come `(tensor([[1, 2], [5, 6]]), tensor([[3, 4], [7, 8]]))`. It does both elementwise conversion and collation.

However, when the dataset is an iterable, we want to only do conversion by considering the entire input as a single data sample. E.g., `[([1, 2], [3, 4]), ([5, 6], [7, 8)]` should be come a single `tensor([[[1, 2], [5, 6]], [[3, 4], [7, 8]]])`.

SsnL commented on torch/utils/data/_utils/pin_memory.py:
```diff
@@ -35,23 +35,23 @@ def _pin_memory_loop(in_queue, out_queue, device_id, done_event):
+        return [pin_memory_data(sample) for sample in data]
     else:
-        return batch
+        return data
```
Because data loader can be used to load individual samples as well, and this PR provides better support for that use case. `batch` seems to assume that the data is always batched.

SsnL commented on torch/utils/data/_utils/signal_handling.py:
```diff
@@ -33,8 +33,9 @@
-    _error_if_any_worker_fails, _set_worker_signal_handlers
+from torch._C import _set_worker_pids, _remove_worker_pid, \
+    _remove_all_worker_pids, _error_if_any_worker_fails, \
+    _set_worker_signal_handlers
```
nah, they are used as `signal_handling. _remove_all_worker_pids` in `dataloader.py`. I will add a comment.

SsnL commented on torch/utils/data/_utils/worker.py:
```diff
@@ -75,9 +115,31 @@ def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed,
         if init_fn is not None:
             init_fn(worker_id)
 
+        from torch.utils.data import DataLoaderMode
```
Sure :) We can always publicize it later if there's reasonable requirement.

SsnL commented on torch/utils/data/_utils/worker.py:
```diff
@@ -87,23 +149,34 @@ def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed,
+                        data = convert_fn(next(dataset_iter))
+                    except StopIteration:
+                        data = IterableDatasetStopIteration(worker_id)
+                        iteration_end = True  # don't waste resource in future iter
```
`try..except..`. The other important use of this flag is to prevent sending more than one `IterableDatasetStopIteration` to main proc. I will add this in the comment.

SsnL commented on torch/utils/data/_utils/worker.py:
```diff
@@ -87,23 +149,34 @@ def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed,
-                del samples
+                data_queue.put((idx, data))
+                del data
+            del idx, index, r
```
will do :)

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -21,6 +24,12 @@
+class DataLoaderMode(object):
+    Map = "map_dataset"
+    MapWithBatchedRead = "map_dataset_with_batched_read"
+    Iterable = "iterable_dataset"
```
Sure. I'll make them ints!

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -34,6 +43,10 @@ class DataLoader(object):
             at every epoch (default: ``False``).
         sampler (Sampler, optional): defines the strategy to draw samples from
             the dataset. If specified, ``shuffle`` must be False.
+        drop_last (bool, optional): set to ``True`` to drop the last incomplete batch,
```
I moved it because it is only used together with `batch_size`, `shuffle` and `sampler`. Grouping them together seemed to make sense to me, but maybe I should respect the arg order more.

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -121,13 +144,91 @@ def __setattr__(self, attr, val):
+
+    def __len__(self):
+        if self.mode == DataLoaderMode.Iterable:
+            raise NotImplementedError
```
Sure.

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -70,47 +79,61 @@ class DataLoader(object):
-                    sampler = SequentialSampler(dataset)
-            batch_sampler = BatchSampler(sampler, batch_size, drop_last)
-
-        self.sampler = sampler
```
Good point. Will use old name.

SsnL commented on torch/utils/data/dataloader.py:
```diff
@@ -121,13 +144,91 @@ def __setattr__(self, attr, val):
+            if self.mode == DataLoaderMode.Map:
+                data = self.convert_fn(self.dataset[index])
+            else:
+                # mode == DataLoaderMode.MapWithBatchedRead:
```
Because previous there is no easy way to to unbatch loading from a map-like dataset. One has to use `batch_size=1` and provide a custom collate_fn which is just `lambda x: x[0]`. That feels counterintuitive to me. With all other changes in this PR, this was a simple addition, so I made it.

SsnL replied:
Another reason is to provide better support for the case where the sampler provides a list of indices, and the dataset does bulk loading.

SsnL commented:
@apaszke There is no new multiprocessing iterator design. I just refactored it into two classes, each handling single or multi proc loading. The logic is exactly the same.

ssnl committed:
Address some comments; Extended timeout

ssnl committed:
Add py2 compatibility for TestProperExitIterableDataset; address some
comments

ssnl committed:
Greatly expand the doc for IterableDataset
Add ChainDataset (the analog of ConcatDataset but for IterableDataset)
and tests
Add doc entries
Make torch.utils.data.* usable without another import after import torch
Make IterableDataset return NotImplemented in __len__ so fallback of
some functions work

ssnl committed:
Single map mode when batch_size is None

SsnL commented:
@pytorchbot retest this please

ssnl committed:
Fix worker err detection on get timeout

ssnl committed:
test_proper_exit: worker_error now only errors in one worker for better testing purpose

ssnl committed:
Rename methods and better format

ssnl committed:
use new cpu half stack

RicCu commented on torch/utils/data/dataset.py:
```diff
@@ -44,9 +138,9 @@ def __len__(self):
-    Purpose: useful to assemble different existing datasets, possibly
+    r"""Dataset to concatenate multiple datasets.
+
+    This classuseful to assemble different existing datasets, possibly
```
nit: typo
```suggestion
    This class is useful to assemble different existing datasets, possibly
```

RicCu commented on torch/utils/data/dataset.py:
```diff
@@ -87,8 +183,36 @@ def cummulative_sizes(self):
+class ChainDataset(IterableDataset):
+    r"""Dataset to chain multiple :class:`IterableDataset` s.
+
+    This classuseful to assemble different existing dataset streamss, possibly
```
nit: typos
```suggestion
    This class is useful to assemble different existing dataset streams, possibly
```

concretevitamin commented:
@SsnL please take a look at https://github.com/pytorch/pytorch/issues/17451 to see if this proposal can support that use case!  /cc @soumith

SsnL commented:
@concretevitamin Yes this would support that use case.

concretevitamin commented:
@SsnL great, looking forward to seeing this PR getting merged!

ssnl committed:
Merge remote-tracking branch 'upstream' into dl_no_sampler

ssnl committed:
fix typos

ssnl committed:
only use cvt_fn for map batch; wip docs

ssnl committed:
fix default_collate

ssnl committed:
fix memory pinning

ssnl committed:
fix custom memory pinning

ssnl committed:
fix submodule changes

ssnl committed:
Write more docs

ssnl committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into dl_no_sampler

ssnl committed:
revert hub.rst changes

ssnl committed:
Improve docs

ssnl committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into dl_no_sampler

ssnl committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into dl_no_sampler

ssnl committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into dl_no_sampler

ssnl committed:
slightly improve test_iterable_dataset

ssnl committed:
More docs

ssnl committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into dl_no_sampler

faroit commented on docs/source/data.rst:
```diff
@@ -2,11 +2,219 @@ torch.utils.data
+Loading from an iterable-style dataset
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+
```
It would be nice to have an example here so that people not familiar with the `utils.data` internals can still discuss the API and usage scenarios.

SsnL replied:
Thanks for the suggestion. I agree. This part isn't finished yet. :)

faroit replied:
no problem, take the time, its an important feature ;-)

I just thought it would be nice to do these docs in an early stage so that "we users" could comment on the API

ssnl committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into dl_no_sampler

ssnl committed:
test_iterable_dataset: better comparison using sorted lists

ssnl committed:
Use subprocess to test_iterable_dataset in multiprocessing

ssnl committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into dl_no_sampler

ssnl committed:
use namedtuple to make flake8 happy

ssnl committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into dl_no_sampler

ssnl committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into dl_no_sampler

ssnl committed:
enable stderr for test_iterable_dataset

ssnl committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into dl_no_sampler

ssnl committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into dl_no_sampler

ssnl committed:
Merge branch 'master' of https://github.com/pytorch/pytorch into dl_no_sampler

ssnl committed:
Add shutdown_worker to prevent code duplication

ssnl committed:
debug ci prints

ssnl committed:
Detect error in init

SsnL commented:
close in favor of https://github.com/pytorch/pytorch/pull/19228
